2024.01.08
=============
# javascript

## async/await
非同期処理の順番を元にして実行する関数
・async
keywordを関数の前に付けると、疲れた関数はPromise関数を返却する
async関数は非同期を処理する、その中でawaitkeywordを使用してPromiseOjbectの解決を待つ。
・await
async関数内でしか使えない、非同期処理が終わるまで待つ
awaitを利用して順番を制御する
await関数はPromiseObjectが解決されるまで、関数の実行止めって、該当Promise結果を返却する
```javascript
export default{
	data() {
		return {
			result; null
		};
	},
methods: {
	async fetchData(){
		try {
			const responce = await this.fechDataFromServer();
			// fechDataFromServerが成功した時
			this.result = responce.data;
		} catch (error) {
			// fechDataFromServerが失敗した時
				console.error("", error);
		}
	},
fetchDataFromServer() {
	return fetch();
}
}
};
```
## promise
・Promise Objectは resolve, rejectを返却するJavaScript非同期処理
・非同期処理が終わった後完了、失敗を表すJavascriptのObject
→resolveはpromiseObjectが成功した時呼び出される
→rejectはpromiseObjectが失敗した時呼び出される
```javascript
const fetchData = new Promise((resolve, reject) => {
	const data = fetchSomeData();
		if (data){
			resolve(data);
		}else{
			reject("error");
		}
)
-> try,catch를 사용하면 이렇게도가능함
const fetchData = new Promise((resolve, reject) => {
	try{
	const data = fetchSomeData();
		if (data){
			resolve(data);
		}else{
			reject("error");
		}
	}catch(error){
		reject(error);
}
)
```